### ISSUE TITLE
max() crashes on long arrays

### ISSUE THREAD

**Issue by sgcrews** [processing/p5.js#6101]
### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [X] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Build Process
- [ ] Unit Testing
- [ ] Internalization
- [ ] Friendly Errors
- [ ] Other (specify if possible)

### p5.js version

1.5.0

### Web browser and version

111.

### Operating System

Windows 10 22H2

### Steps:
1. Create a simple 1D array of floats of length > ~ 126,000
2. Pass the array as argument to max()
3. "Maximum call stack size exceeded" error is reported

### Snippet:

```js

// Paste your code here :)
  const m = new Array(126000);
  let myMax = max(m);
```

**Comment by aditya-shrivastavv** [comment]
This problem has nothing to do specifically with Math.max.
Function.prototype.apply can only receive array of limited length as its second argument.

```javascript
p5.prototype.max = function(...args) {
  if (args[0] instanceof Array) {
    return Math.max.apply(null, args[0]);
  } else {
    return Math.max.apply(null, args);
  }
};
```
On locally testing, it in Chrome using:

```js
function limit(l) {
  var x = []; x.length = l;
  (function (){}).apply(null, x);
}
```

Locally, limit(l) crashed exactly with l = 124980.
means the the array of size 126000 you want to pass is already exceeding the limit.

I don't think their is a direct solution to it, other than implementing our own

[mdn on apply](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply#using_apply_and_built-in_functions)

**Comment by sgcrews** [comment]
Is the limitation because max takes a recursive approach and the number of recursions is limited? 
I am working around it in my own script by breaking the array into chunks that are small enough to use max() on, but this is somewhat tedious.

**Comment by limzykenneth** [comment]
The limitation is because `Math.max()` (or rather any function in JS) can only accept a limited number of arguments when being called. p5.js `max()` while able to accept an array, it is passing that array to `Math.max()` as individual arguments. One way to get around this limit is to use the following

```js
const arr = [1, 2, 3];
const max = arr.reduce((a, b) => Math.max(a, b), -Infinity);
```

as recommended by MDN. However, this will likely be slower (maybe even significantly slower) than if you have a smaller list of numbers.

**Comment by davepagurek** [comment]
We could maybe fall back to that method if the user passes in an array over some threshold length. Wouldn't be able to do anything about it if they use multiple parameters, but it'd be something.

**Comment by limzykenneth** [comment]
@davepagurek There is the underscore/lodash method of looping through the whole array and not using Math.max which will be faster than reduce/Math.max but I'm not sure how that will benchmark against just Math.max for smaller list of numbers.

**Comment by aditya-shrivastavv** [comment]
This [stack overflow](https://stackoverflow.com/questions/18308700/chrome-how-to-solve-maximum-call-stack-size-exceeded-errors-on-math-max-apply) thread is completely relevant to the issue.

**Comment by pmarsh-scottlogic** [comment]
I'm looking to make a contribution, so I've gone and read through those forums and tried to understand the problem.

### Approaches
From what I've seen, the obvious ways to find the max of a list of numbers are as follows:

**Apply**
`return Math.max.apply(null, arr);`

**Iterate**
```
let max = -Infinity;
for (let i = 0; i < arr.size; i++) {
max = arr[i] > max ? arr[i] : max;
}
return max;
```

**Spread**
`return Math.max(...arr)`

**Reduce**
`return arr.reduce((a, b) => Math.max(a, b), -Infinity);`

We are currently using **apply** which is problematic because it fills up the stack. Apparently in V8, each element is added to the stack individually before any operations on elements are carried out.

### Benchmarking
I ran each method through jsbench.me to see which was fastest for different array sizes. Here's the benchmark project https://jsbench.me/anlgxgqdm2/2

**small array, size 10**
![image](https://user-images.githubusercontent.com/118171430/234548683-c5a4af39-5cec-4ea2-955a-a82b8b6534ce.png)
The Iterate method wind by a mile on small arrays.

**big array, size 100,000**
![image](https://user-images.githubusercontent.com/118171430/234558108-f215778b-bf23-4c88-bfff-6725540518ee.png)
Iterate method wins again.

**really big array, size 1,000,000
![image](https://user-images.githubusercontent.com/118171430/234559228-18f7e6fd-6227-45d6-a34e-d1d2b22080b5.png)
Iterate method gets the hat-trick! Reduce runs, but slowly and the other two methods hit the same stack error.

I'm running these on Chrome, Version 112.0.5615.138 (Official Build) (64-bit).

for Chrome:
size| reduce | spread | apply | iterate
--- | --- | --- | --- | ---
10 | 150M ops/s | 19M ops/s | 125M ops/s | **2.1B ops/s**
100,000 | 478 ops/s | 515 ops/s | 433 ops/s | **633M ops/s**
1,000,000 | 72 ops/s | stack error | stack error | **1.1B ops/s**

**Comment by davepagurek** [comment]
Thanks for making such detailed benchmarks! This seems like a pretty convincing reason to go with our own `max()` implementation using a loop under the hood.

**Comment by pmarsh-scottlogic** [comment]
I'll make a PR quick then :)

**Comment by davepagurek** [comment]
This looks good to me! @limzykenneth does this seem ok?

**Comment by limzykenneth** [comment]
I believe the consistency and high performance of the iterate benchmark is due to the js engine's JIT which will typically cache loop results when run repeatedly, the actual performance difference might not be that significant. I don't remember if browser nowadays allow JIT to be disabled but that may give a more real world performance indication.

`reduce` I've expected it to be the slowest because it needs to call a callback function repeatedly.

Spread and `apply` are essentially the same thing and I think when the code is transpiled through babel they will likely end up having the same or similar implementation. There is also a possibility to combine for loop with spread/apply by processing the array in batches of say around 120000 which is the case I'm not sure about whether it will outperform just one simple loop through everything.

**Comment by pmarsh-scottlogic** [comment]
Okay I've refactored the test to try to eliminate any clever caching that could go on in the background, simply by generating a fresh array for each iteration of the test. On inspection it seems to give much more useful results [I acknowledge that the array generation is now included in the benchmarking time, but if we assume that each array generation takes the same time, then we can still identify a ranking across our max methods]. Let me know if you see any flaws with this benchmarking approach.

Here's what it looks like now:
https://jsbench.me/anlgxgqdm2
![image](https://user-images.githubusercontent.com/118171430/235158576-e0e240ff-e3e6-44a8-baae-c631523cc835.png)

And here's the results over several browsers:
#### Chrome:
<html>
<body>
<google-sheets-html-origin><style type="text/css"></style>

array size | reduce | spread | apply | iterate
-- | -- | -- | -- | --
10 | **13M** | 7.6M | 10M | 12M
10^5 | 120 | 229 | 362 | **447**
10^7 | 4.6 | stack error | stack error | **6**


</body>
</html>

#### Firefox:
<html>
<body>
<google-sheets-html-origin><style type="text/css"></style>

array size | reduce | spread | apply | iterate
-- | -- | -- | -- | --
10 | 9M | 21M | 22M | **24M**
10^5 | 962 | 1.2K | 1.2K | **2.4K**
10^7 | 5.5 | RangeError | RangeError | **7.5**


</body>
</html>

#### Edge:
<html>
<body>
<google-sheets-html-origin><style type="text/css"></style>

array size | reduce | spread | apply | iterate
-- | -- | -- | -- | --
10 | **5.8M** | 3.9M | 5.3M | 5M
10^5 | 231 | 256 | 258 | **391**
10^7 | 2.7 | RangeError | RangeError | **3.8**


</body>
</html>

So overall, my judgement is that iterate is still the best method. It is fastest, or at least not significantly slower, than the other methods across every array size. Plus, in my view, a simpler implementation without splitting the array into chunks is more readable.

**Comment by pmarsh-scottlogic** [comment]
I'm attempting to make the change but having lots of issue with the line endings! (using windows and vscode)
Bear with me 😅

**Comment by davepagurek** [comment]
@pmarsh-scottlogic depending on what your issues are, maybe the fix in this PR might help you? https://github.com/processing/p5.js/pull/6124

**Comment by pmarsh-scottlogic** [comment]
> @pmarsh-scottlogic depending on what your issues are, maybe the fix in this PR might help you? #6124

That's the badger! Thanks for bringing that to my attention - great timing :)
